include/onnxruntime/core/session/onnxruntime_c_api.h

```diff
@@ -959,6 +967,11 @@ struct OrtApi {
    */
   void(ORT_API_CALL* ClearBoundInputs)(_Inout_ OrtIoBinding* binding_ptr) NO_EXCEPTION ORT_ALL_ARGS_NONNULL;
   void(ORT_API_CALL* ClearBoundOutputs)(_Inout_ OrtIoBinding* binding_ptr) NO_EXCEPTION ORT_ALL_ARGS_NONNULL;
+
+  // Supports CPU device only.
+  ORT_API2_STATUS(CreateAllocatorForSharing, _In_ const OrtMemoryInfo* mem_info, _In_ const OrtArenaCfg* arena_cfg,
+                  _Outptr_ OrtAllocator** out);
+  ORT_API2_STATUS(RegisterSharedAllocator, _Inout_ OrtEnv* env, _Inout_ OrtAllocator* allocator);
```

Comment:
Presumably the allocator needs to live longer than all the sessions? Is it destroyed in the same way (given it's been registered with the env)?

Reply:
Do we need separate create + register or can we do automatic registration inside 'create' so there's no question about lifetime?

---
In reply to: [471182831](https://github.com/microsoft/onnxruntime/pull/4813#discussion_r471182831) [](ancestors = 471182831)

Reply:
Added these as 2 separate APIs because I want to leave the option to the user to implement his own allocator (using the OrtAllocator interface) and register it with ORT using the Register API. Alternative is to have another Register API (different name since overloads are not allowed) that takes OrtMemoryInfo and OrtArenaCfg as parameters.